In [1]:
print("Book Recommender System")

Book Recommender System


Books Data Link: https://www.kaggle.com/datasets/ra4u12/bookrecommendation 

In [18]:
# Import libraries
import pandas as pd
import numpy as np

### Notes on `pd.read_csv` Parameters

- **`sep=";"`**  
  The CSV file uses **semicolon `;`** as the column separator instead of the default comma `,`.  
  Pandas uses this to correctly split each line into columns.

- **`error_bad_lines=False`**  
  Tells Pandas to **skip rows that cause parsing errors** instead of crashing.  
  For example, if a row has the wrong number of columns, it will be ignored.

- **`encoding='latin-1'`**  
  Specifies the **text encoding** of the CSV file.  
  Many datasets from Europe use `latin-1` instead of the default `utf-8`.  
  This avoids errors with special characters (like accented letters).


In [ ]:
# Read the book data (BX-Books.csv) -> 1st Data
books = pd.read_csv("../data/BX-Books.csv",  # Data path
                    sep=";", # Columns separated by semicolons 
                    error_bad_lines=False, # Skip rows with parsing errors (
                    encoding='latin-1') # File uses Latin-1 encoding to handle special characters

/home/arju/miniconda3/envs/books/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3552: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)
b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping 

In [20]:
# Show top 3 data(rows)
books.head(3)

ISBN            Book-Title           Book-Author Year-Of-Publication  \
0  0195153448   Classical Mythology    Mark P. O. Morford                2002   
1  0002005018          Clara Callan  Richard Bruce Wright                2001   
2  0060973129  Decision in Normandy          Carlo D'Este                1991   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2          HarperPerennial  http://images.amazon.com/images/P/0060973129.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...

In [21]:
# Access the 'Image-URL-S' value of the 11th book (index 10)
books.iloc[10]['Image-URL-S']


'http://images.amazon.com/images/P/0771074670.01.THUMBZZZ.jpg'

In [22]:
books.shape  # Get the shape of the DataFrame (number of rows and columns)

(271360, 8)

- `271360` → Number of rows (each row represents a book)
- `8` → Number of columns (each column is a feature, like Book-Title, Book-Author,Publisher, Image-URL-S, etc.)


In [23]:
books.columns  # Get the column names of the DataFrame

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [31]:
# Rename the columns for better readability
books.rename(
    columns={
        'Book-Title': 'title',
        'Book-Author': 'author',
        'Year-Of-Publication': 'publication',
        'Publisher': 'publisher',
        'Image-URL-S': 'img_url_s',
        'Image-URL-M': 'img_url_m',
        'Image-URL-L': 'img_url' # We will work on 'Large' Size image for our future need, So, we rename it from img_url_l' to img_url
    },
    inplace=True
)

In [32]:
books.head()

ISBN                                              title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author publication                   publisher  \
0    Mark P. O. Morford        2002     Oxford University Press   
1  Richard Bruce Wright        2001       HarperFlamingo Canada   
2          Carlo D'Este        1991             HarperPerennial   
3      Gina Bari Kolata        1999        Farrar Straus Giroux   
4       E. J. W. Barber        1999  W. W. Norton &amp; Company   

                                             img_url  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           img_url_m  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                             img_url  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [34]:
# Read the book data (BX-Users.csv) -> 2nd Data
users  =pd.read_csv("../data/BX-Users.csv",
                    sep=";",
                    error_bad_lines=False,
                    encoding='latin-1')

/home/arju/miniconda3/envs/books/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3552: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [35]:
users.head(3)

User-ID                         Location   Age
0        1               nyc, new york, usa   NaN
1        2        stockton, california, usa  18.0
2        3  moscow, yukon territory, russia   NaN

In [36]:
users.columns

Index(['User-ID', 'Location', 'Age'], dtype='object')

In [37]:
users.shape

(278858, 3)

- `278858` → Number of users (rows)
- `3` → Number of features/columns (User-ID, Location, Age)

In [42]:
# Rename the columns to lowercase for consistency
users.rename(columns={
    'User-ID':'user_id',
    "Location":'location',
    "Age":'age'
}, inplace=True)

In [43]:
users.head()

user_id                            location   age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [44]:
# Read the book data (BX-Book-Ratings.csv) -> 3nd Data
from os import error


ratings = pd.read_csv("../data/BX-Book-Ratings.csv",
                      sep=';',
                      error_bad_lines=False,
                      encoding='latin-1')


/home/arju/miniconda3/envs/books/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3552: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [45]:
ratings.head(3)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0

In [46]:
ratings.shape

(1149780, 3)

- `1,149,780` → Number of rows (each row is a rating given by a user).
- `3` → Number of columns (User-ID, ISBN, Book-Rating).


In [47]:
ratings.columns

Index(['User-ID', 'ISBN', 'Book-Rating'], dtype='object')

In [ ]:
# Rename the columns for easier access
ratings.rename(columns={
    'User-ID':'user_id',
    'Book-Rating':'rating'

}, inplace=True)

In [49]:
ratings.head(3)

user_id        ISBN  rating
0   276725  034545104X       0
1   276726  0155061224       5
2   276727  0446520802       0